### PR TITLE
correctly update ContentLength for uncompressed response body

### PR DESCRIPTION
**Pull request by ruitianzhong**
fix the bug that is found when reproducing in issue #491 .

When Content-Encoding is `gzip`, the uncompressed response body length is different from the compressed one. So the Content-Length need to be updated to stay consistent with the uncompressed body length.

More detail in https://github.com/gojue/ecapture/issues/491#issuecomment-1972430408

### CODE REVIEW

**Review by cfc4n — COMMENTED** [review]
Can you add a unit test?

**Comment by ruitianzhong** [comment]
* first I add three *.bin files. In particular, 952253698002000.bin file is response body which is compress with `gzip` to detect the bug we fix in this PR.
To see its content, run the following content
```bash
cat 952253698002000.bin | gunzip
```
* Also error detection is added in `TestEventProcessor_Serve()` function to explicitly detect error through log message.

With all steps above, it can detect all related bugs on version that is only added unit test related code with the bug not fixed.

```bash
=== RUN   TestEventProcessor_Serve
    processor_test.go:81: 2024/03/01 20:00:21 [http response] DumpResponse error: unexpected EOF
    processor_test.go:81: 2024/03/01 20:00:21 [http response] DumpResponse error: http: ContentLength=1145 with Body length 2443
    processor_test.go:89: some errors occurred
--- FAIL: TestEventProcessor_Serve (3.00s)
```
It can be verified through my demo branch https://github.com/ruitianzhong/ecapture/tree/demo

* Furthermore, with the added unit test, one more error is found during testing (i.e., DumpResponse error: unexpected EOF). This is because of the wrong value for Content-Length (74332, actually too large) in 952282712204824.bin. 

* Finally, with all the bug fixes applied, all the unit tests are passed.

File: pkg/event_processor/testdata/952282712204824.bin [hunk 1]
```diff
@@ -2,7 +2,7 @@ HTTP/1.1 200 OK
 Server: nginx/1.14.0 (Ubuntu)
 Date: Wed, 29 Jun 2022 09:44:35 GMT
 Content-Type: text/html; charset=UTF-8
-Content-Length: 74332
+Content-Length: 20124
```
Comment (cfc4n) [inline pkg/event_processor/testdata/952282712204824.bin:5]: Why has the number of bytes increased here?

File: pkg/event_processor/processor_test.go [hunk 2]
```diff
@@ -68,12 +71,23 @@ func TestEventProcessor_Serve(t *testing.T) {
...
-	select {
-	case <-tick.C:
-	}
+	<-tick.C
+
 	err = ep.Close()
+	lines = strings.Split(buf.String(), "\n")
+	ok := true
+	for _, line := range lines {
+		if strings.Contains(strings.ToLower(line), "dump") {
```
Comment (cfc4n) [inline pkg/event_processor/processor_test.go:80]: Why is a for loop needed to determine whether the string contains `dump`? Which test file contains the string `dump`?

**Review by cfc4n — COMMENTED** [review]
PTAL

File: pkg/event_processor/processor_test.go (hunk 2 above)
Comment (ruitianzhong) [reply pkg/event_processor/processor_test.go:80]: In the unit test here, we store all the output from the logger and try to extract string that indicate error. In this case, things like `DumpResponse error: unexpected EOF` is our target. If the any line contains it the this test fails and the exact message is printed out.

File: pkg/event_processor/testdata/952282712204824.bin (hunk 1 above)
Comment (ruitianzhong) [reply pkg/event_processor/testdata/952282712204824.bin:5]: Actually, the number of bytes decreased here (from 74332 -> 20124). 

The total response body length for the following `*.bin` file is obviously less than 74332 which will cause the error `DumpError: Unexpected EOF`. This problem has existed before and could be verified through https://github.com/gojue/ecapture/actions/runs/8094294516/job/22118690803#step:8:226
    
```json
{"DataType":0,"Timestamp":952282712204824,"Pid":469953,"Tid":469953,"DataLen":4096,"Comm":[99,117,114,108,0,0,0,0,0,0,0,0,0,0,0,0],"Fd":5,"Version":771}
{"DataType":0,"Timestamp":952282715334491,"Pid":469953,"Tid":469953,"DataLen":4096,"Comm":[99,117,114,108,0,0,0,0,0,0,0,0,0,0,0,0],"Fd":5,"Version":771}
{"DataType":0,"Timestamp":952282731469385,"Pid":469953,"Tid":469953,"DataLen":4096,"Comm":[99,117,114,108,0,0,0,0,0,0,0,0,0,0,0,0],"Fd":5,"Version":771}
{"DataType":0,"Timestamp":952282736644923,"Pid":469953,"Tid":469953,"DataLen":4096,"Comm":[99,117,114,108,0,0,0,0,0,0,0,0,0,0,0,0],"Fd":5,"Version":771}
{"DataType":0,"Timestamp":952282737583175,"Pid":469953,"Tid":469953,"DataLen":4096,"Comm":[99,117,114,108,0,0,0,0,0,0,0,0,0,0,0,0],"Fd":5,"Version":771}
```

**Comment by ruitianzhong** [comment]
**Further Explanation**

On previous commit(before this PR):

When running:
```bash
wget -d --header 'Accept-Encoding: gzip' https://1.1.1.1
```
It works well and no `DumpError` occurred.

However, when running:
```bash
wget -d --header 'Accept-Encoding: gzip' https://www.baidu.com
```
`DumpError` occurred(mentioned in #491).

After further investigation, I found that it's because of the response structure.

for baidu, its response header is like
```
HTTP/1.1 200 OK
Content-Encoding: gzip
Content-Length: 1145
Server: bfe
Date: Fri, 01 Mar 2024 08:24:27 GMT
Content-Type: text/html;charset=UTF-8
```
for 1.1.1.1, its response header is like:
`HTTP/1.1 200 OK
Content-Type: text/plain
Transfer-Encoding: chunked``
```
Obviously, Content-Length is different. According to the [HTTP/1.1 standard](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Transfer-Encoding), with `Transfer-Encoding` set, Content-Length is not set and in go it is zero, according to comment in http package comment:

```go
// ContentLength records the length of the associated content. The
	// value -1 indicates that the length is unknown. Unless Request.Method
	// is "HEAD", values >= 0 indicate that the given number of bytes may
	// be read from Body.
	ContentLength [int64](https://pkg.go.dev/builtin#int64)
```
So `DumpResponse()` work well under this condition after response body is uncompressed.

However, when the Content-Length is set like first example, if we just uncompressed the response body without adjusting the Content-Length, error would be returned from `DumpResponse`. 

Just to give  more context info here.

Thanks!

**Comment by cfc4n** [comment]
LGTM, thanks.

In fact, when a larger string is called with SSL_Write, it will be split into multiple fragments and SSL_Write will be called multiple times. Moreover, there are concurrent scenarios in the upper-level business logic. 
Therefore, there is an issue of disorder in the order of HOOK results for SSL_Write. The current solution cannot guarantee a perfect resolution.

Therefore, a better solution is to use the `-m pcap` mode.

**Review by cfc4n — APPROVED** [review]
merged

**Comment by ruitianzhong** [comment]
I agree with you. There exists a semantic gap between eCapture and the application which means that eCapture might not fully understand the application-specific behavior (like parsing different interleaving protocol). When that really happen, some errors occur which is expected though

**Pull request merged**